### PR TITLE
MatrixStore powered API endpoints

### LINKED ISSUE #1519: MatrixStore: second sprint
There are two high level aims here:
1. Integrate the MatrixStore build step into the import pipeline.
2. Switch out the Analyse API backend to be powered by the MatrixStore.

## 1. Integrate the MatrixStore build step into the import pipeline

- [ ] The filename argument to the `matrixstore_build` command should become optional. Omitting the filename should result in the file being created in the "appropriate" place, with that to be decided as per questions below.
- [ ] A step should be added to the import pipeline which runs `matrixstore_build`. This should be after the prescribing and practice statistics data is loaded into BigQuery, but doesn't need to wait for any data to be loaded into Postgres.

### Things to consider:

This step will add several hours to the import time. Do we want to consider consider removing the "import warning" banner, sending the Slack message etc before this step has completed? How awkward is it to do this?

Where should the SQLite file live? Initially I thought `PIPELINE_DATA_BASE_DIR` was the obvious choice, but this is on the root partition which isn't enormous. So perhaps it needs to go on one of the much larger partitions in `/mnt`.

How should the file be named? My preference based on similar past use cases would be something like:
`prescribing_<DATETIME_OF_BUILD>_<LATEST_PRESCRIBING_MONTH>_<HASH_OF_FILE_CONTENTS>.sqlite`
Including the build date gives a useful natural ordering, and having the hash is useful as part of a key for caching purposes.

## 2. Switch out the Analyse API backend to be powered by the MatrixStore

- [ ] Add some state which tells us what the currently live MatrixStore file is. Probably a new database table, but see discussion below
- [ ] Add a management command which updates this state to make a new version of the file live.
- [ ] Add a step to the import pipeline which calls this command after successfully building the new file. (In future we will want to be able to review data before pushing live, but for now it's better that it automatically stays in sync with the data in Postgres.)
 - [ ] Replace the endpoints used by the Analyse page with MatrixStore powered ones. These are
    - [ ] `spending_by_practice` which takes:
      *  an optional list of BNF code prefixes (or things which can be converted to that)
       * an optional list of practice codes (or CCG codes which get translated to practice codes)
       * an optional date
    - [ ] `spending_by_ccg` which takes:
      *  an optional list of BNF code prefixes (or things which can be converted to that)
       * an optional list of CCG codes
- [ ] Use a sample of historical requests from the log to confirm that the old and new endpoints return the same results
- [ ] Using the same sample, compare timings of old and new endpoints to confirm performance improvement
- [ ] Stress test the new endpoints using both nasty queries and high concurrency using e.g. apache bench
- [ ] Confirm that the Trends pages still work correctly as these use the same spending endpoints as the Analyse form

### Things to consider:

Both endpoints can accept an empty list of BNF prefixes and will sum over all prescribing. The interface won't allow you to do this for the numerator, but it will allow it for the denominator so we need to support it. This is currently quite an expensive query to do in the MatrixStore as it involves loading all the data which, from memory, can take in the order of 45 seconds.

One option would be to precompute global totals matrices. This feels a bit icky as part of the point of the MatrixStore was to avoid all the complexity and inflexibility of pre-computation. But summing over the next level down in the BNF hierarchy (chapter, is it?) is sufficiently fast that it may be we only need this one kludge.

We need some state which tells us which file to use for live data. There are various ways to do this, including using symlinks, but I think my preference would be to store this in Postgres. The simplest implementation would be a single row in a dedicated table which stores the filename. There would also be the option for keeping a log of when each new version was deployed.

Wherever this state ends up being we will want some mechanism for updating it. Now that we have the shiny new ebmbot, a reasonable option might be to do this via a management command and then have an ebmbot task that calls the management command. This saves us building a web UI for it.

### PR DESCRIPTION
**Pull request by evansd**
This uses the MatrixStore to power the backend for the following `api/1.0/` endpoints: `spending`, `spending_by_practice`, `spending_by_ccg` and `org_details`.

The only intended user-visible change here is that the restriction on mixing levels of the BNF hierarchy has been removed.

Two unintended but hopefully acceptable changes are:

 * The `org_details` endpoint no longer returns results for organisations with a zero list size. Previously, the endpoint would omit results if there was no corresponding practice statistics row, but would return zero valued rows when it found them. The matrix data structure can't distinguish zero-valued from missing entries and so we omit all zero-valued entries.

 * We now store costs in pence as integers, rather than in pounds as floats. This makes no difference to net costs which only ever had two decimal places in any case. But "actual" costs were stored to a much higher precision -- a precision which was utterly spurious given that these are estimated values. This does mean that when summing over many values a difference accumulates between the results under the old and new systems. But this difference rarely amounts to more than 0.05% of the total, and where it does the absolute difference is always less than 20p.

Closes #1519

### CODE REVIEW

**Comment by evansd** [comment]
Before merging this needs to be deployed to staging and some load testing done.

Also note, there is one other area when the new API responses differ from the old and that is any request involving this [duplicate data](https://github.com/ebmdatalab/openprescribing/issues/976), where the old API just returns incorrect results.

File: openprescribing/api/views_spending.py [hunk 1]
```diff
@@ -483,82 +503,118 @@ def tariff(request, format=None):
...
-    data = utils.execute_query(query, params)
-    return Response(data)
+        date_offsets = [(date, db.date_offsets[date])]
+    else:
+        date_offsets = sorted(db.date_offsets.items())
+    # Yield entries for each organisation on each date
+    for date, col_offset in date_offsets:
+        for org, row_offset in org_offsets:
+            index = (row_offset, col_offset)
+            items = items_matrix[index]
```
[the remark this answers is not in the document]
Comment (evansd) [reply openprescribing/api/views_spending.py]: Yes, I think we definitely will want something that handles this for us. But I deliberately didn't write anything now as I want the dust to settle on this before we start baking in one particular abstraction. I think the time to do this will be when we start changing the API (or creating a new API) to support new features.

File: openprescribing/matrixstore/grouper.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import numpy
+
+
+class Grouper(object):
+    """
+    The `Grouper` class provides a callable which sums together groups of
+    matrix rows. We generally use it for aggregating practice level data into
+    organisations that consist of groups of practices like CCGs or STPs.
+    """
```
[the remark this answers is not in the document]
Comment (evansd) [reply openprescribing/matrixstore/grouper.py]: No, "grouper" is almost certainly not the right name but I'm not 100% sold on "summer" either, and not just because it's homographic for the season. The SQL operation it replaces is `SUM ... GROUP BY` and that's tricky to nounify. I wonder about calling the class `SumByGroup` and then calling instances of it 'sum_by_org' so we'd have code that looked like this:
```python
sum_by_org = sum_by(org_type)
items_matrix = sum_by_org(items_matrix)
```

Or maybe we could stick with "group by" in the variable name as it feels obvious to me that in this context we're summing within the groups:
```python
group_by_org = group_by(org_type)
items_matrix = group_by_org(items_matrix)
```

I don't think this is pure bikeshedding: it's worth getting right if we can.

[the remark this answers is not in the document]
Comment (sebbacon) [reply openprescribing/matrixstore/grouper.py]: How about something like this:
```python
offsets_to_group = [(0, 'even'), (1, 'odd'), (2, 'even')]

matrix = np.array([
  [1, 2, 3, 4],
  [2, 3, 4, 5],
  [3, 4, 5, 6],
])

grouped = MatrixWithGroup(
    matrix,
    group_axis='rows',
    offsets_to_group=offsets_to_group)

# Included `output_groups` so it's explicit how the returned matrix is sorted row-wise
grouped.apply(np.sum, output_groups=['even', 'odd'])
```

File: openprescribing/matrixstore/tests/test_grouper.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+from itertools import product
+import random
+
+from django.test import SimpleTestCase
+
+from matrixstore.grouper import Grouper
+from matrixstore.matrix_ops import finalise_matrix, sparse_matrix
+
+
+class TestGrouper(SimpleTestCase):
```
Comment (sebbacon) [inline openprescribing/matrixstore/tests/test_grouper.py]: I wonder if this is a case where classic fixtures would be better than the compute-everything-with-a-different-code-path route?

The current approach has the advantage of programmatically covering all the edge cases, but the disadvantage that it's hard for a new reader to validate it's doing what we expect. This makes sense for things like measures because it's hard however you do it, but fixtures in this case are simple and easy to validate by reading, e.g.

```python
input_data = np.array([
    [1, 2, 3],
    [0, 1, 2],
    [0, 0, 0]]
)
groups = [(0, 'even'), (1, 'odd'), (2, 'even')]
expected = np.array([
    [1, 2, 3],
    [0, 1, 2]]
)
```

File: openprescribing/matrixstore/grouper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (evansd) [reply openprescribing/matrixstore/grouper.py]: I wonder if it would be enough just to rename `Grouper` to `RowGrouper` and `__call__` to either `sum` or, if we want to be even more explicit, `sum_by_group`.

So the calling code would look like this:
```python
group_by_org = get_row_grouper(org_type)
items_matrix = group_by_org.sum(items_matrix)
```

File: openprescribing/matrixstore/tests/test_grouper.py (hunk 3 above)
Comment (evansd) [reply openprescribing/matrixstore/tests/test_grouper.py]: I think the answer is just that we want both!

File: openprescribing/matrixstore/grouper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (inglesp) [reply openprescribing/matrixstore/grouper.py]: I quite like this.

[the remark this answers is not in the document]
Comment (sebbacon) [reply openprescribing/matrixstore/grouper.py]: Yes, I think this is better and doesn't require changing much code.

I still quite like the idea of `group_by_org.apply(np.sum, items_matrix)` because it makes it very clear what's going on, but perhaps YAGNI in terms of any other aggregate function ever needing to be used.  And not particularly attached to it...

**Comment by sebbacon** [comment]
Notes about performance [here](https://gist.github.com/d25e7b56a3c5ee6b1f4ddf369254b4b2).

TLDR it's fine - doesn't scale well but we don't really care - overall a tad slower, but more performant for queries matching a subset of practices and/or chemicals and less performant for summary stastistics at paragraph level and higher.

As these are used for denominators, we will want to think about this sooner rather than later -- presumably via caching.

There is an opportunity to investigate IO to improve performance. I've not got very far, but  significant amounts of CPU time are spent waiting for disk IO under SQLite.   Playing with page cache settings appears to introduce pathological behaviour in some edge cases, but I've not had time to be very methodical about this.

In short we should probably cease investigation as this is Good Enough but I am intrigued about the IO thing.

**Comment by evansd** [comment]
Thanks @sebbacon, that's interesting and slightly disappointing, particularly given the discrepancy with the performance I got locally.

But then looking at the CPU specs, it's clear that my laptop is actually considerably more powerful than the server:
```
$ lscpu | grep -P 'CPU|Thread'
CPU op-mode(s):      32-bit, 64-bit
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
CPU family:          6
Model name:          Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
CPU MHz:             500.000
CPU max MHz:         3400.0000
CPU min MHz:         400.0000
NUMA node0 CPU(s):   0-3

$ ssh largeweb2.ebmdatalab.net lscpu | grep -P 'CPU|Thread'
CPU op-mode(s):        32-bit, 64-bit
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
CPU family:            6
Model name:            Intel(R) Xeon(R) CPU E5-2650L v3 @ 1.80GHz
CPU MHz:               1799.998
NUMA node0 CPU(s):     0-3
```

We've already discussed how we could get considerably more bang for the buck if we used dedicated hardware from e.g. OVH. So maybe that's an option longer term if performance starts to become more of a problem.

### Page cache

One small clarification: I didn't realised "page cache" meant something specific in SQLite-land; I was just referring to the kernel page cache and the fact that if you access the same block of unchanging data on disk the kernel will end up caching it in RAM.

In fact, chasing this down a bit more it transpires that SQLite has two entirely different IO systems and the one I assumed it used is not in fact the default. I think we want to be using [mmapped IO](https://www.sqlite.org/mmap.html), which we can enable by issuing a pragma statement:
```
PRAGMA mmap_size=<SOME_LARGE_NUMBER_OF_BYTES>;
```
If we set that to larger than the size of the file that SQlite will map the entire file into memory and then rely on the kernel page cache to read the data off disk when needed, rather than implementing its own cache on top of this.

Additionally, because the cache is shared between processes we ought to get much better concurrency performance as each gunicorn process doesn't need to reload the data.

If the test instance is still up I think it would be worth having a go at setting this and seeing what difference it makes.

**Comment by sebbacon** [comment]
Ah, I didn't notice that option, good find.

While it's true that our CPU may limit things a lot, it looked to me that most of the CPU time was in `WAIT`, so I suspect your laptop was faster because SSD more than anything.

Running some tests now, let's see... though per above I should draw a line even if inconclusive...

**Comment by sebbacon** [comment]
This gave a marginal speedup, but I can still see loads of disk IO.

Most of the memory is assigned to the cache:
```
root@largeweb2-test:/webapps/openprescribing/openprescribing# free -h
              total        used        free      shared  buff/cache   available
Mem:            15G        4.4G        151M         10G         11G        760M
Swap:            0B          0B          0B
```

We allow postgres a lot of memory, not sure if that might be interacting:

```
#------------------------------------------------------------------------------
# RESOURCE USAGE (except WAL)
#------------------------------------------------------------------------------

# - Memory -

shared_buffers = 10GB           # min 128kB
work_mem =250MB                 # min 64kB; per parallel gather worker
maintenance_work_mem = 5GB              # min 1MB
dynamic_shared_memory_type = posix      # the default is the first option
                                        # supported by the operating system:
                                        #   posix
                                        #   sysv
                                        #   windows
                                        #   mmap
                                        # use none to disable dynamic shared memory

```

I tuned those figures down and restarted postgres, and re-ran the tests:

```
# free -h
              total        used        free      shared  buff/cache   available
Mem:            15G        4.9G        7.9G         94M        2.8G         10G
Swap:            0B          0B          0B
```

However, although it seemed like there was a *lot* less IO, performance was still the same, so perhaps CPU sys/user time is indeed the main limiter.

I have written up the test setup [here](https://gist.github.com/sebbacon/e75c597791e4bc7c1e9f98846a4e153a) for future reference. Will turn off the droplet now but not delete it for a few days in case you want to play @evansd  - if not, please could you delete it?

Regarding CPU performance - if we moved to Google Cloud there are more options regarding CPU too.

**Comment by evansd** [comment]
OK, thanks. We've got 32GB on the production server so hopefully it won't require any Postgres reconfiguration.
